### ISSUE TITLE
[$750] [MEDIUM] Update Recents in Participants page to reflect the type of action taken

### ISSUE THREAD

**Issue by mountiny** [Expensify/App#34227]
### Problem

Coming from [this discussion](https://expensify.slack.com/archives/C02MW39LT9N/p1704820424041429?thread_ts=1704753651.553139&cid=C02MW39LT9N).

When taking some action like requesting money or assigning a task, the participants page does not take the type of the action into consideration when creating the list of Recents. We list out the people/ workspace chats we have interacted with the latest even if we have for example never requested money from them.

### Solution

Update the Global create participants page Recents list to showrecents based on the type of action user is taking.

So if they are taking:
- creating scan request
- creating distance request
- creating manual request
- creating task
- creating split bill

for each of these, try to pick the recents such we list first the chats where the user has most recently done such action. If no data to do this optimization is available, default to the current behaviour
<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01b6cc63ba2f30aad7</li>
        <li>Upwork Job ID: 1749580420206743552</li>
        <li>Last Price Increase: 2024-02-05</li>
<li>Automatic offers: </li>
<ul>
<li>rojiphil | Contributor | 28122004</li>
</ul></li>
    </ul>
</details>

**Comment by shawnborton** [comment]
cc @trjExpensify - we were chatting about this one today

**Comment by muttmuure** [comment]
@mountiny do we need to get this into a Wave or VIP?

**Comment by trjExpensify** [comment]
It's on the wave6 project board already, but yeah.. needs progressing. Vit can maybe confirm if it needs to stay internal or not.

**Comment by mountiny** [comment]
I think we can try this Externally. We would need to look through the trasactions and then identified the reports which we requested that given type of action latest from

**Comment by mountiny** [comment]
Not sure from your message, flow Global Crete:
- I choose Manual request
- I choose scan and take picture
- on the participant page the recents should list the latest in choronological order chats I have requested money with scan request.

If there was no scan request before, just keep the current logic

If there would be only one or less than current recents, just show that number of chats (so there could be only 1 in recents) this is up for discussion though

Same goes for other types. For tasks this might be hard to do for requests it should be possible as we should have all the transactions locally

**Comment by rojiphil** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

We list out the people/ workspace chats we have interacted with the latest even if we have for example never requested money from them. The same problem can be extended to `tasks` as well. We want to solve this problem by displaying the
recent list in the participant page such that recent reports based on the type of action taken by the user are displayed.

### What is the root cause of that problem?

This is a new feature.

The scope is as follows:
1) Show recent reports based on the following type of actions chosen by user
a) Scan Request creation b) Distance request creation c) Manual request creation d) Split bill creation e) Task creation
2) For each of the above, try picking recents for the specific type of action. If no such results are available, show the recents based on current behaviour which is the default one.

### What changes do you think we should make in order to solve the problem?

For money request action involving various types of requests, the participant list is displayed through `IOURequestStepParticipants`. This page makes use of `MoneyRequestParticipantsSelector` [here](https://github.com/Expensify/App/blob/a1fe671e94f69a694e2847a6b74685270149e991/src/pages/iou/request/step/IOURequestStepParticipants.js#L90C1-L96C15). 

So, the first thing to do is to determine the type of money request action by relying on `iouRequestType` and number of `participants` like this [here](https://github.com/Expensify/App/blob/cc64de9ff6cb556927821ab84c9ddf491be91437/src/pages/iou/request/MoneyTemporaryForRefactorRequestParticipantsSelector.js#L85): 

    const actionType = iouType === CONST.IOU.TYPE.REQUEST && participants.length > 0 ? CONST.IOU.TYPE.SPLIT : iouRequestType;

Further, we can pass `actionType` to `getFilteredOptions` [here](https://github.com/Expensify/App/blob/cc64de9ff6cb556927821ab84c9ddf491be91437/src/pages/iou/request/MoneyTemporaryForRefactorRequestParticipantsSelector.js#L99C1-L124C11). 

To implement this, we can add a helper method in `TransactionUtils` called `getTransactionsByRequestType` which will help us to get the transactions based on the action type. This helper method implementation can be like this: 

```
function getTransactionsByRequestType(iouRequestType?: string): Array<OnyxEntry<Transaction>> {
    return Object.values(allTransactions ?? {})
        .filter((transaction): transaction is Transaction => transaction != null && (iouRequestType === CONST.IOU.TYPE.SPLIT ? isSplitRequest(transaction) : getRequestType(transaction) ===  iouRequestType))
        .sort((transactionA, transactionB) => {
            const transactionATime = new Date(transactionA?.created);
            const transactionBTime = new Date(transactionB?.created);
            return transactionATime.valueOf() - transactionBTime.valueOf();
        })
        .reverse();
}
```

Also, since there is no helper method for identifying `split` request, we can add a helper for this by checking against `comment.source` like this:

```
function isSplitRequest(transaction: Transaction): boolean {
    return Boolean(transaction?.comment?.source === CONST.IOU.TYPE.SPLIT);
}
```

Now, we can retrieve the recent transactions based on the action type like this [here](https://github.com/Expensify/App/blob/cc64de9ff6cb556927821ab84c9ddf491be91437/src/libs/OptionsListUtils.js#L1482):

```
    const isMoneyRequestActionType = actionType === CONST.IOU.REQUEST_TYPE.MANUAL || actionType === CONST.IOU.REQUEST_TYPE.SCAN || actionType === CONST.IOU.REQUEST_TYPE.DISTANCE || actionType === CONST.IOU.TYPE.SPLIT;
        let recentTransactions = [];
        if (isMoneyRequestActionType) {
            recentTransactions = TransactionUtils.getRecentTransactionsByRequestType(actionType );
        }
```

And, then, populate a new array `recentReportOptionsByAction` based on action type like this [here](https://github.com/Expensify/App/blob/cc64de9ff6cb556927821ab84c9ddf491be91437/src/libs/OptionsListUtils.js#L1529):

            if (isMoneyRequestActionType && recentTransactions.some((transaction) => transaction.reportID === String(reportOption.iouReportID))) {
                recentReportOptionsByAction.push(reportOption);
            }

            // Stop adding options to the recentReports by action array when we reach the maxRecentReportsToShow value
            if (recentReportOptionsByAction.length > 0 && recentReportOptionsByAction.length === maxRecentReportsToShow) {
                break;
            }

Similarly for `tasks`, we can pass `CONST.REPORT.TYPE.TASK` as `actionType` in `TaskAssigneeSelectorModal` [here](https://github.com/Expensify/App/blob/cc64de9ff6cb556927821ab84c9ddf491be91437/src/pages/tasks/TaskAssigneeSelectorModal.js#L92C1-L108C11). Further, we can push the `reportOption` for tasks to `recentReportOptionsByAction` like this:

```
            if (actionType == CONST.REPORT.TYPE.TASK && reportOption.isTaskReport) {
                console.log("PROCESSING actionType[TASK], reportID["+reportOption.reportID+"]");
                recentReportOptionsByAction.push(reportOption);
            }
```

Finally, when available, we can display recent reports based on `recentReportOptionsByAction` or fallback to the existing `recentReportOptions` [here](https://github.com/Expensify/App/blob/cc64de9ff6cb556927821ab84c9ddf491be91437/src/libs/OptionsListUtils.js#L1635) like this:

`         recentReports: recentReportOptionsByAction.length > 0 ? recentReportOptionsByAction : recentReportOptions,`

### What alternative solutions did you explore? (Optional)

**Comment by mountiny** [comment]
cc @shawnborton would you want to assign this one to yourself to oversee the correct logic implementation?

**Comment by shawnborton** [comment]
I would prefer that someone from BZ or the project team oversees it, but happy to help test and guide implementation.

**Comment by mountiny** [comment]
@eVoloshchak can you review the proposal above?

**Comment by eVoloshchak** [comment]
@rojiphil's [proposal](https://github.com/Expensify/App/issues/34227#issuecomment-1908444762) looks good to me!

> iouRequestType?: string

We can use `iouRequestType?: ValueOf<typeof CONST.IOU.REQUEST_TYPE>`

🎀👀🎀 C+ reviewed!

**Comment by mountiny** [comment]
Melvin, what a prankster

**Comment by rojiphil** [comment]
Thank you for the assignment here. 
I will start work on this today and provide an update over the weekend.

**Comment by eVoloshchak** [comment]
Not overdue, PR is being drafted by @rojiphil

**Comment by rojiphil** [comment]
Update: Will share PR for review by tomorrow this time.

**Comment by rojiphil** [comment]
@mountiny @eVoloshchak 
Do you think we can consider an increase in the bounty here?
The reason why I think so is that I notice quite a lot of thoughtful work involved due to the sheer number of test cases involved here.

**Comment by mountiny** [comment]
Increased to 750, noting that a regression still cuts the price in half though so please be thorough as possible in testing. Thank you

**Comment by eVoloshchak** [comment]
This should be open, the PR is still in review ([recent changes broke the logic a bit](https://github.com/Expensify/App/pull/35431#issuecomment-2065629123))

**Comment by trjExpensify** [comment]
Agreed!

**Comment by muttmuure** [comment]
Still waiting on linked PR to be merged

**Comment by muttmuure** [comment]
PR still open

**Comment by mountiny** [comment]
I am still torn on this implementation. It seems to be bringing really complex changes to the App to try to deduce the recent target accounts for various actions, but its not reliable (as we might not have the report actions available after signing in) and the changes are hard to follow.

I am now more inclined towards a solution with backend change similar to the quick action button if we really want this feature. this would simplify the logic and code. 

Whenever user takes some action from the global create flows, we would update an NVP with the target reportIDs, it would basically be same as the QAB NVP, but we would keep rolling history of the 5 recent reports for any given flow. 

@rojiphil @eVoloshchak you have been battling with the implementation but it seems like there has been numerous small issues with your solution and the code is hard to follow, what is your opinion?

**Comment by rojiphil** [comment]
Thinking of this a little further, this is what I could I arrive at:
1. We can take the advantage of quick actions [here](https://github.com/Expensify/App/blob/807c945ef8edabe86cc62f16c866aa754db8750b/src/CONST.ts#L620C1-L632C7) which would map to the mentioned flow:
```
        REQUEST_MANUAL==> MANUAL
        REQUEST_SCAN==> SCAN
        REQUEST_DISTANCE==> DISTANCE
        SPLIT_MANUAL==> SPLIT	
        SPLIT_SCAN==> SPLIT	
        SPLIT_DISTANCE==> SPLIT	
        TRACK_MANUAL==> MANUAL	
        TRACK_SCAN==> SCAN	
        TRACK_DISTANCE==> DISTANCE	
        ASSIGN_TASK==> TASK	
        SEND_MONEY==> MANUAL
```
2. The following Onyx entries will be added:
```
	np_recent_manual_requests
	nvp_recent_scan_requests
	nvp_recent_distance_requests
	nvp_recent_split_requests
	nap_recent_tasks
```
Each of these will hold the flow specific array of reportIDs with rolling history of the 5 recent reports

3. For every quick action that reaches BE, the BE will push the report ID in the flow specific `nvp_recent_*` list and send/push the updated entries to FE.

4. The FE will also optimistically update the flow specific `nvp_recent_*` Onyx entries by pushing the reportID.

5. While displaying recent list, the FE will use the flow specific Onyx entries. If not present, the default one will be displayed as shown currently.

6. We also need to remove the redundant `reportID` from the recent list in certain cases like leaving a group chat. Also, BE need to update the recent list in such cases and push the updated entries back to FE.
Also, there would be cases where a report is not present in Onyx but is present in recent list. Such reports would also be not shown in recent list. So we might end up with less than 5 entries in recent list on removal of such reports. Hope this is fine.  

@mountiny Is this what we are aiming for? Or are there any changes required?

**Comment by mountiny** [comment]
I think we would need to hold internal discussion about this, currently focusing on higher priority tasks though

**Comment by muttmuure** [comment]
No update

**Comment by muttmuure** [comment]
Not overdue

**Comment by mountiny** [comment]
This task showed up to be quite of a rabbit hole. A proper solution would require a backend changes too something similar along the lines of QAB but we did not really prioritise this. Instead we are trying out this change https://github.com/Expensify/App/issues/46683 and lets see if it will be enough from the UX perspective.

I think we could close this issue out for focus @trjExpensify what do you think? We can always come back to this if we want to prioritize creating a proper fix for this. We would of course pay some reward for the effort

**Comment by trjExpensify** [comment]
Yep, works for me!

**Comment by rojiphil** [comment]
> We would of course pay some reward for the effort

@mountiny Thanks for mentioning this. I also think this deserves to be rewarded as quite a lot of effort went into it from my side. Would $500 sound fair here instead of the actual price of $750? 

Reasonings:
1. The PR itself needed a lot of work to cover about 7 different test cases as mentioned in the OP [here](https://github.com/Expensify/App/pull/35431#issue-2108589328)
2. Quite a lot of review cycles occurred here which included verification via [Timing tests](https://github.com/Expensify/App/pull/35431#issuecomment-1952974405) apart from regular re-reviews that spanned for almost about 2 months.

**Comment by mountiny** [comment]
@rojiphil Yes I think thats valid. I would go for 50% of the planned amount so $375, how does that sound? @eVoloshchak

**Comment by eVoloshchak** [comment]
@mountiny, yeah, a 50% payout seems fair to me

**Comment by rojiphil** [comment]
> I would go for 50% of the planned amount so $375, how does that sound?

Thanks @mountiny. 
Maybe we can make this issue `daily` to close on this sooner. 
cc @muttmuure

**Comment by muttmuure** [comment]
Handling

**Comment by muttmuure** [comment]
$375 - @rojiphil 
$375 - @eVoloshchak

**Comment by muttmuure** [comment]
@eVoloshchak please submit your request in New Expensify!

**Comment by JmillsExpensify** [comment]
$375 approved for @eVoloshchak
